lib/rpmug.c

```diff
@@ -1,16 +1,79 @@
 #include "system.h"
 
-#include <pthread.h>
-#include <pwd.h>
-#include <grp.h>
-#include <netdb.h>
+#include <errno.h>
 #include <rpm/rpmlog.h>
 #include <rpm/rpmstring.h>
 
 #include "lib/misc.h"
 #include "lib/rpmug.h"
 #include "debug.h"
 
+#define PWFILE "/etc/passwd"
+#define GRPFILE "/etc/group"
+
+/*
+ * Lookup an arbitrary field based on contents of another in a ':' delimited
+ * file, such as /etc/passwd or /etc/group.
+ */
+static int lookup_field(const char *path, const char *val, int vcol, int rcol,
+			char **ret)
+{
+    int rc = -1; /* assume not found */
+    char *str, buf[BUFSIZ];
+    FILE *f = fopen(path, "r");
+    if (f == NULL) {
+	rpmlog(RPMLOG_ERR, _("failed to open %s for id/name lookup: %s\n"),
+		path, strerror(errno));
+	return rc;
+    }
+
+    while ((str = fgets(buf, sizeof(buf), f)) != NULL) {
+	int nf = vcol > rcol ? vcol : rcol;
+	const char *fields[nf + 1];
+	char *tok, *save = NULL;
+	int col = -1;
+
+	while ((tok = strtok_r(str, ":", &save)) != NULL) {
+	    fields[++col] = tok;
+	    str = NULL;
+	    if (col >= nf)
+		break;
+	}
+
+	if (col >= nf) {
+	    if (rstreq(val, fields[vcol])) {
+		*ret = xstrdup(fields[rcol]);
+		rc = 0;
+	    }
+	}
+    }
+
+    fclose(f);
+
+    return rc;
+}
+
+static int lookup_num(const char *path, const char *val, int vcol, int rcol,
+			long *ret)
+{
+    char *buf = NULL;
+    int rc = lookup_field(path, val, vcol, rcol, &buf);
+    if (rc == 0) {
+	long num = atol(buf);
```

Comment:
Any reason to not use strtol() here?  In the error case here, doesn't atol() give you 0 which may or may not be fine.  But strtol() could catch errors for reporting here.

Reply:
Heh, the reason to use atol() over strtol() is "I'll add error handling later" (as in, when the rest of it works) and then promptly forgot :laughing: A fair point and will fix. Erring to root might not be the safest behavior here...

Reply:
Oh I've been there.  I will definitely add error checking later.  :)

This one stood out to me because of the 0 returned for errors which would be UID or GID 0.

Reply:
Now with error checking :sweat_smile: